worldedit-bukkit/src/main/java/com/sk89q/bukkit/util/ClassSourceValidator.java

```diff
@@ -0,0 +1,128 @@
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.bukkit.util;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.Maps;
+import org.bukkit.plugin.Plugin;
+
+import java.security.CodeSource;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Validates that certain specified classes came from the same source as
+ * a plugin.
+ */
+public class ClassSourceValidator {
+
+    private static final Logger log = Logger.getLogger(ClassSourceValidator.class.getCanonicalName());
+    private static final String separatorLine = Strings.repeat("*", 46);
+
+    private final Plugin plugin;
+    @Nullable
+    private final CodeSource expectedCodeSource;
+
+    /**
+     * Create a new instance.
+     *
+     * @param plugin The plugin
+     */
+    public ClassSourceValidator(Plugin plugin) {
+        checkNotNull(plugin, "plugin");
+        this.plugin = plugin;
+        this.expectedCodeSource = plugin.getClass().getProtectionDomain().getCodeSource();
+    }
+
+    /**
+     * Return a map of classes that been loaded from a different source.
+     *
+     * @param classes A list of classes to check
+     * @return The results
+     */
+    public Map<Class<?>, CodeSource> findMismatches(List<Class<?>> classes) {
+        checkNotNull(classes, "classes");
+
+        Map<Class<?>, CodeSource> mismatches = Maps.newHashMap();
+
+        if (expectedCodeSource != null) {
+            for (Class<?> testClass : classes) {
+                CodeSource testSource = testClass.getProtectionDomain().getCodeSource();
+                if (!expectedCodeSource.equals(testSource)) {
+                    mismatches.put(testClass, testSource);
+                }
+            }
+        }
+
+        return mismatches;
+    }
+
+    /**
+     * Reports classes that have come from a different source.
+     *
+     * <p>The warning is emitted to the log.</p>
+     *
+     * @param classes The list of classes to check
+     */
+    public void reportMismatches(List<Class<?>> classes) {
+        if (Boolean.getBoolean("enginehub.disable.class.validation")) {
+            return;
+        }
+        Map<Class<?>, CodeSource> mismatches = findMismatches(classes);
+
+        if (!mismatches.isEmpty()) {
+            StringBuilder builder = new StringBuilder("\n");
+
+            builder.append(separatorLine).append("\n");
+            builder.append("** /!\\    SEVERE WARNING    /!\\\n");
+            builder.append("** \n");
+            builder.append("** A plugin developer has copied and pasted a portion of \n");
```

Comment:
```suggestion
            builder.append("** A plugin developer has included a portion of \n");
```
I don't think we should assume copy/paste, shading is likely too (and while still technically copy/paste, it's not quite the same).